File: src/openjarvis/learning/agents/skill_discovery.py
"""Skill discovery -- mine recurring tool sequences from traces."""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from typing import Any, Dict, List, Tuple


@dataclass(slots=True)
class DiscoveredSkill:
    """A skill discovered from trace analysis."""
    name: str
    description: str
    tool_sequence: List[str]  # ordered tool names
    frequency: int  # how often this sequence appeared
    avg_outcome: float  # average outcome score
    example_inputs: List[str] = field(default_factory=list)


class SkillDiscovery:
    """Mine recurring tool sequences from trace data to auto-generate skills.

    Analyzes TraceStore data for patterns like:
    - "web_search -> file_write" (research-then-save)
    - "file_read -> calculator -> file_write" (read-compute-save)

    When a sequence appears >= min_frequency times with positive outcomes,
    it's surfaced as a DiscoveredSkill that can be registered.
    """

    def __init__(
        self,
        *,
        min_frequency: int = 3,
        min_sequence_length: int = 2,
        max_sequence_length: int = 4,
        min_outcome: float = 0.5,
    ) -> None:
        self._min_freq = min_frequency
        self._min_len = min_sequence_length
        self._max_len = max_sequence_length
        self._min_outcome = min_outcome
        self._discovered: List[DiscoveredSkill] = []

    def analyze_traces(self, traces: List[Any]) -> List[DiscoveredSkill]:
        """Analyze a list of traces for recurring tool sequences.

        Parameters
        ----------
        traces:
            List of Trace objects (or dicts with 'steps' and 'outcome' keys).
            Each trace should have steps with 'step_type' and 'tool_name'.

        Returns
        -------
        List of DiscoveredSkill objects meeting frequency and outcome thresholds.
        """
        # Extract tool sequences from traces
        sequence_data: Dict[Tuple[str, ...], List[float]] = defaultdict(list)
        sequence_inputs: Dict[Tuple[str, ...], List[str]] = defaultdict(list)

        for trace in traces:
            tool_calls = self._extract_tool_sequence(trace)
            outcome = self._extract_outcome(trace)
            query = self._extract_query(trace)

            if len(tool_calls) < self._min_len:
                continue

            # Generate all subsequences of valid length
            upper = min(self._max_len + 1, len(tool_calls) + 1)
            for length in range(self._min_len, upper):
                for start in range(len(tool_calls) - length + 1):
                    seq = tuple(tool_calls[start:start + length])
                    sequence_data[seq].append(outcome)
                    if query and len(sequence_inputs[seq]) < 3:
                        sequence_inputs[seq].append(query)

        # Filter by frequency and outcome
        discovered = []
        for seq, outcomes in sequence_data.items():
            freq = len(outcomes)
            avg_outcome = sum(outcomes) / len(outcomes) if outcomes else 0.0

            if freq >= self._min_freq and avg_outcome >= self._min_outcome:
                name = "_".join(seq)
                desc = f"Auto-discovered skill: {' -> '.join(seq)} (seen {freq} times)"
                discovered.append(DiscoveredSkill(
                    name=name,
                    description=desc,
                    tool_sequence=list(seq),
                    frequency=freq,
                    avg_outcome=avg_outcome,
                    example_inputs=sequence_inputs.get(seq, []),
                ))

        # Sort by frequency * outcome (quality score)
        discovered.sort(key=lambda s: s.frequency * s.avg_outcome, reverse=True)
        self._discovered = discovered
        return discovered

    def _extract_tool_sequence(self, trace: Any) -> List[str]:
        """Extract ordered list of tool names from a trace."""
        if isinstance(trace, dict):
            steps = trace.get("steps", [])
        elif hasattr(trace, "steps"):
            steps = trace.steps
        else:
            return []

        tools = []
        for step in steps:
            if isinstance(step, dict):
                if step.get("step_type") == "tool_call":
                    name = step.get("tool_name", step.get("name", ""))
                    if name:
                        tools.append(name)
            elif hasattr(step, "step_type"):
                st = step.step_type
                is_tool = str(st) == "tool_call" or (
                    hasattr(st, "value") and st.value == "tool_call"
                )
                if is_tool:
                    name = getattr(
                        step, "tool_name", getattr(step, "name", ""),
                    )
                    if name:
                        tools.append(name)
        return tools

    def _extract_outcome(self, trace: Any) -> float:
        """Extract outcome score from a trace."""
        if isinstance(trace, dict):
            return float(trace.get("outcome", 0.0))
        return float(getattr(trace, "outcome", 0.0))

    def _extract_query(self, trace: Any) -> str:
        """Extract the original query from a trace."""
        if isinstance(trace, dict):
            return trace.get("query", "")
        return getattr(trace, "query", "")

    @property
    def discovered_skills(self) -> List[DiscoveredSkill]:
        """Return the most recently discovered skills."""
        return list(self._discovered)

    def to_skill_manifests(self) -> List[Dict[str, Any]]:
        """Convert discovered skills to TOML-compatible manifest dicts."""
        manifests = []
        for skill in self._discovered:
            manifests.append({
                "name": skill.name,
                "description": skill.description,
                "steps": [
                    {"tool": tool, "params": {}} for tool in skill.tool_sequence
                ],
                "metadata": {
                    "auto_discovered": True,
                    "frequency": skill.frequency,
                    "avg_outcome": skill.avg_outcome,
                },
            })
        return manifests


__all__ = ["DiscoveredSkill", "SkillDiscovery"]

File: rust/crates/openjarvis-learning/src/skill_discovery.rs
//! Skill discovery — mine recurring tool sequences from traces.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscoveredSkill {
    pub name: String,
    pub description: String,
    pub tool_sequence: Vec<String>,
    pub frequency: usize,
    pub avg_outcome: f64,
    pub example_inputs: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct SkillDiscovery {
    min_freq: usize,
    min_len: usize,
    max_len: usize,
    min_outcome: f64,
    discovered: Vec<DiscoveredSkill>,
}

/// Accumulator for a candidate subsequence during analysis.
struct SeqAccum {
    outcomes: Vec<f64>,
    inputs: Vec<String>,
}

impl SkillDiscovery {
    pub fn new(
        min_frequency: usize,
        min_sequence_length: usize,
        max_sequence_length: usize,
        min_outcome: f64,
    ) -> Self {
        Self {
            min_freq: min_frequency,
            min_len: min_sequence_length,
            max_len: max_sequence_length,
            min_outcome,
            discovered: Vec::new(),
        }
    }

    /// Analyze traces for recurring tool sequences.
    ///
    /// Each trace entry is `(tool_calls, outcome_score, query)`.
    /// Returns a slice of discovered skills meeting all thresholds.
    pub fn analyze(&mut self, traces: &[(Vec<String>, f64, String)]) -> &[DiscoveredSkill] {
        let mut accums: HashMap<Vec<String>, SeqAccum> = HashMap::new();

        for (tool_calls, outcome, query) in traces {
            if tool_calls.len() < self.min_len {
                continue;
            }
            let upper = (self.max_len + 1).min(tool_calls.len() + 1);
            for length in self.min_len..upper {
                for start in 0..=(tool_calls.len() - length) {
                    let seq = tool_calls[start..start + length].to_vec();
                    let acc = accums.entry(seq).or_insert_with(|| SeqAccum {
                        outcomes: Vec::new(),
                        inputs: Vec::new(),
                    });
                    acc.outcomes.push(*outcome);
                    if !query.is_empty() && acc.inputs.len() < 3 {
                        acc.inputs.push(query.clone());
                    }
                }
            }
        }

        let mut discovered: Vec<DiscoveredSkill> = Vec::new();
        for (seq, acc) in accums {
            let freq = acc.outcomes.len();
            let avg = acc.outcomes.iter().sum::<f64>() / freq as f64;
            if freq >= self.min_freq && avg >= self.min_outcome {
                let name = seq.join("_");
                let desc = format!(
                    "Auto-discovered skill: {} (seen {} times)",
                    seq.join(" -> "),
                    freq,
                );
                discovered.push(DiscoveredSkill {
                    name,
                    description: desc,
                    tool_sequence: seq,
                    frequency: freq,
                    avg_outcome: avg,
                    example_inputs: acc.inputs,
                });
            }
        }

        discovered.sort_by(|a, b| {
            let score_a = a.frequency as f64 * a.avg_outcome;
            let score_b = b.frequency as f64 * b.avg_outcome;
            score_b
                .partial_cmp(&score_a)
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        self.discovered = discovered;
        &self.discovered
    }

    pub fn discovered_skills(&self) -> &[DiscoveredSkill] {
        &self.discovered
    }

    /// Convert discovered skills to TOML-compatible manifest values.
    pub fn to_manifests(&self) -> Vec<serde_json::Value> {
        self.discovered
            .iter()
            .map(|skill| {
                serde_json::json!({
                    "name": skill.name,
                    "description": skill.description,
                    "steps": skill.tool_sequence.iter()
                        .map(|t| serde_json::json!({"tool": t, "params": {}}))
                        .collect::<Vec<_>>(),
                    "metadata": {
                        "auto_discovered": true,
                        "frequency": skill.frequency,
                        "avg_outcome": skill.avg_outcome,
                    },
                })
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_traces() -> Vec<(Vec<String>, f64, String)> {
        vec![
            (
                vec!["web_search".into(), "file_write".into()],
                0.8,
                "research topic A".into(),
            ),
            (
                vec!["web_search".into(), "file_write".into()],
                0.9,
                "research topic B".into(),
            ),
            (
                vec!["web_search".into(), "file_write".into()],
                0.7,
                "research topic C".into(),
            ),
            (
                vec![
                    "file_read".into(),
                    "calculator".into(),
                    "file_write".into(),
                ],
                0.85,
                "compute stats".into(),
            ),
        ]
    }

    #[test]
    fn test_discovers_frequent_sequence() {
        let mut sd = SkillDiscovery::new(3, 2, 4, 0.5);
        let traces = sample_traces();
        let skills = sd.analyze(&traces);
        assert!(!skills.is_empty());
        let ws_fw = skills
            .iter()
            .find(|s| s.tool_sequence == vec!["web_search", "file_write"]);
        assert!(ws_fw.is_some(), "should discover web_search -> file_write");
        assert_eq!(ws_fw.unwrap().frequency, 3);
    }

    #[test]
    fn test_filters_below_min_freq() {
        let mut sd = SkillDiscovery::new(10, 2, 4, 0.5);
        let traces = sample_traces();
        let skills = sd.analyze(&traces);
        assert!(skills.is_empty(), "nothing should meet freq >= 10");
    }

    #[test]
    fn test_to_manifests() {
        let mut sd = SkillDiscovery::new(3, 2, 4, 0.5);
        let traces = sample_traces();
        sd.analyze(&traces);
        let manifests = sd.to_manifests();
        assert!(!manifests.is_empty());
        assert!(manifests[0].get("name").is_some());
        assert!(manifests[0].get("steps").is_some());
    }
}
